### PR TITLE
fix: install ESM for Ubuntu 18.04

### PR DESCRIPTION
**Pull request by anujmaheshwari1**
**What type of PR is this?**
This PR uses Ubuntu Advantage to install ESM(Extended Security Maintenance) to Ubuntu 18.04 until 1.27 GA's, after which all of AKS will default to 20.04

Successful VHD build for VHDs: https://msazure.visualstudio.com/CloudNativeCompute/_build/results?buildId=73447595&view=results

image
image

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Requirements**:
- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Special notes for your reviewer**:

**Release note**:
```
none
```

### CODE REVIEW

File: vhdbuilder/packer/pre-install-dependencies.sh [hunk 1]
```diff
@@ -48,11 +48,9 @@ rm -f /etc/cron.daily/logrotate
 
 systemctlEnableAndStart sync-container-logs.service || exit 1
 
-if [[ (${UBUNTU_RELEASE} == "20.04" || ${UBUNTU_RELEASE} == "18.04" || ($OS == $MARINER_OS_NAME && $OS_VERSION == "2.0")) && ${ENABLE_FIPS,,} == "true" ]]; then
-  installFIPS
-elif [[ ${ENABLE_FIPS,,} == "true" ]]; then
-  echo "AKS enables FIPS on Ubuntu 18.04, 20.04 or Mariner 2.0 only, exiting..."
-  exit 1
+if [[ ${UBUNTU_RELEASE} == "18.04" ]]; then
```
Comment (anujmaheshwari1) [inline vhdbuilder/packer/pre-install-dependencies.sh:51]: the very first step for 18.04 SKUs will be installing ESM

File: vhdbuilder/packer/vhd-image-builder-mariner.json [hunk 2]
```diff
@@ -464,7 +469,32 @@
     {
       "type": "shell",
       "inline": [
-        "sudo FEATURE_FLAGS={{user `feature_flags`}} BUILD_NUMBER={{user `build_number`}} BUILD_ID={{user `build_id`}} COMMIT={{user `commit`}} HYPERV_GENERATION={{user `hyperv_generation`}} CONTAINER_RUNTIME={{user `container_runtime`}} TELEPORTD_PLUGIN_DOWNLOAD_URL={{user `teleportd_plugin_download_url`}} ENABLE_FIPS={{user `enable_fips`}} SGX_INSTALL={{user `sgx_install`}} IMG_SKU={{user `img_sku`}} /bin/bash -ux /home/packer/pre-install-dependencies.sh",
```
Comment (anujmaheshwari1) [inline vhdbuilder/packer/vhd-image-builder-mariner.json:467]: @cameronmeissner when you originally added multiple reboots 5 months ago, Mariner FIPS was not supported. we do build Mariner FIPS SKU now, hence separated pre-install, install-fips and install-deps with a reboot in between each

**Comment by anujmaheshwari1** [comment]
made the reboot behavior consistent across all VHDs, even those that dont install FIPS/ESM.
install ESM (only for 1804) -> reboot -> install FIPS dependencies -> reboot -> install-dependencies -> reboot -> post-install dependencies

File: vhdbuilder/packer/vhd-image-builder-arm64-gen2.json [hunk 3]
```diff
@@ -510,6 +515,19 @@
       "skip_clean": true,
       "pause_after": "60s"
     },
+    {
+      "type": "shell",
+      "inline": [
+        "sudo FEATURE_FLAGS={{user `feature_flags`}} BUILD_NUMBER={{user `build_number`}} BUILD_ID={{user `build_id`}} COMMIT={{user `commit`}} HYPERV_GENERATION={{user `hyperv_generation`}} CONTAINER_RUNTIME={{user `container_runtime`}} TELEPORTD_PLUGIN_DOWNLOAD_URL={{user `teleportd_plugin_download_url`}} ENABLE_FIPS={{user `enable_fips`}} SGX_INSTALL={{user `sgx_install`}} IMG_SKU={{user `img_sku`}} /bin/bash -ux /home/packer/install-fips-dependencies.sh"
+      ]
+    },
+    {
```
Comment (cameronmeissner) [inline vhdbuilder/packer/vhd-image-builder-arm64-gen2.json:524]: ahh I see now why you wanted to avoid this - this'll add a fair bit of latency to non-FIPS builds for no reason

Comment (cameronmeissner) [reply vhdbuilder/packer/vhd-image-builder-arm64-gen2.json:524]: could potentially create new template(s) for FIPS specifically, though that has disadvantages of its own...

wonder what @alexeldeib thinks

File: vhdbuilder/scripts/linux/ubuntu/tool_installs_ubuntu.sh [hunk 4]
```diff
@@ -190,3 +181,37 @@ relinkResolvConf() {
...
+    # Run apt get update to refresh repo list
+    # Run apt dist get upgrade to install packages/kernels
+    apt_get_update || exit $ERR_APT_UPDATE_TIMEOUT
+    apt_get_dist_upgrade || exit $ERR_APT_DIST_UPGRADE_TIMEOUT    
+    
+    # 'ua status' for logging
+    ua status
+
+    echo "detaching ua..."
+    retrycmd_if_failure 5 10 120 printf "y\nN" | ua detach || $ERR_UA_DETACH
```
[the remark this answers is not in the document]
Comment (cameronmeissner) [reply vhdbuilder/scripts/linux/ubuntu/tool_installs_ubuntu.sh:201]: I see it already exists above so most likely fine

[the remark this answers is not in the document]
Comment (anujmaheshwari1) [reply vhdbuilder/scripts/linux/ubuntu/tool_installs_ubuntu.sh:201]: oh yes, FIPS has been doing it just this way for quite some time now. https://github.com/Azure/AgentBaker/blob/aad498d3f57944f91d7a0871d3d95fa84b509b06/vhdbuilder/scripts/linux/ubuntu/tool_installs_ubuntu.sh#L166

**Review by cameronmeissner — APPROVED** [review]

File: vhdbuilder/packer/vhd-image-builder-arm64-gen2.json (hunk 3 above)
Comment (alexeldeib) [reply vhdbuilder/packer/vhd-image-builder-arm64-gen2.json:524]: +1 to https://github.com/Azure/AgentBaker/pull/3205#discussion_r1196780672

avoid the separate reboot. do it inside the script. have the script no-op if it's not fips.

File: vhdbuilder/packer/install-fips-dependencies.sh [hunk 5]
```diff
@@ -0,0 +1,28 @@
+#!/bin/bash
+OS=$(sort -r /etc/*-release | gawk 'match($0, /^(ID_LIKE=(coreos)|ID=(.*))$/, a) { print toupper(a[2] a[3]); exit }')
```
[the remark this answers is not in the document]
Comment (anujmaheshwari1) [reply vhdbuilder/packer/install-fips-dependencies.sh]: removed the file altogether

File: vhdbuilder/packer/vhd-image-builder-arm64-gen2.json (hunk 3 above)
Comment (anujmaheshwari1) [reply vhdbuilder/packer/vhd-image-builder-arm64-gen2.json]: did the UA install in pre-install, back to the original number of reboots

File: vhdbuilder/packer/install-fips-dependencies.sh [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+source /home/packer/provision_source.sh
+source /home/packer/provision_source_distro.sh
+source /home/packer/tool_installs.sh
+source /home/packer/tool_installs_distro.sh
+source /home/packer/packer_source.sh
+
+CPU_ARCH=$(getCPUArch)  #amd64 or arm64
+VHD_LOGS_FILEPATH=/opt/azure/vhd-install.complete
+
+if [[ (${UBUNTU_RELEASE} == "20.04" || ${UBUNTU_RELEASE} == "18.04" || ($OS == $MARINER_OS_NAME && $OS_VERSION == "2.0")) && ${ENABLE_FIPS,,} == "true" ]]; then
```
[the remark this answers is not in the document]
Comment (anujmaheshwari1) [reply vhdbuilder/packer/install-fips-dependencies.sh]: redundant.

**Comment by anujmaheshwari1** [comment]
@alexeldeib I let this dist upgrade stay as is https://github.com/Azure/AgentBaker/blob/aad498d3f57944f91d7a0871d3d95fa84b509b06/vhdbuilder/packer/install-dependencies.sh#LL47C1-L55C3, do you think we can remove it and also remove the reboot after install deps?

**Review by cameronmeissner — APPROVED** [review]

File: vhdbuilder/packer/pre-install-dependencies.sh [hunk 7]
```diff
@@ -48,11 +48,47 @@ rm -f /etc/cron.daily/logrotate
...
+# First handle Mariner + FIPS
+if [[ ${OS} == ${MARINER_OS_NAME} ]]; then
+  if [[ ${ENABLE_FIPS,,} == "true" ]]; then
+    # This is FIPS install for Mariner and has nothing to do with Ubuntu Advantage
+    echo "Install FIPS for Mariner SKU"
+    installFIPS
+  fi
+else
+  # Handle FIPS and ESM for Ubuntu
+  useUA=$( [[ "${UBUNTU_RELEASE}" = "18.04" ]] || [[ "${ENABLE_FIPS,,}" = "true" ]] && echo "true" || echo "false" )
```
Comment (alexeldeib) [inline vhdbuilder/packer/pre-install-dependencies.sh:60]: nit: I think this is actually more chars than just doing it inline :D

Comment (alexeldeib) [reply vhdbuilder/packer/pre-install-dependencies.sh:60]: wait also what is this echo...

File: vhdbuilder/packer/pre-install-dependencies.sh [hunk 8]
```diff
@@ -48,11 +48,47 @@ rm -f /etc/cron.daily/logrotate
...
+  if [[ "${useUA}" == "true" ]]; then
+    # useUA = true implies the SKU is either FIPS Enabled or 1804
+    autoAttachUA
+  fi
```
Comment (alexeldeib) [inline vhdbuilder/packer/pre-install-dependencies.sh:64]:
```suggestion
  if [[ "${UBUNTU_RELEASE}" == "18.04" ]] || [[ "${ENABLE_FIPS,,}" == "true" ]]; then
    autoAttachUA
  fi
```

File: vhdbuilder/packer/pre-install-dependencies.sh (hunk 7 above)
Comment (anujmaheshwari1) [reply vhdbuilder/packer/pre-install-dependencies.sh:60]: oops I was testing the one liner

File: vhdbuilder/scripts/linux/ubuntu/tool_installs_ubuntu.sh [hunk 9]
```diff
@@ -150,22 +150,37 @@ installFIPS() {
         fi
     done
 
+    echo "enabling ua fips-updates..."
+    retrycmd_if_failure 5 10 1200 echo y | ua enable fips-updates || exit $ERR_UA_ENABLE_FIPS
+}
+
+relinkResolvConf() {
```
Comment (alexeldeib) [inline vhdbuilder/scripts/linux/ubuntu/tool_installs_ubuntu.sh:157]: unnecessary diff?

File: vhdbuilder/packer/pre-install-dependencies.sh (hunk 7 above)
Comment (anujmaheshwari1) [reply vhdbuilder/packer/pre-install-dependencies.sh:60]: I was gonna use useUA twice, one to autoattach and once to detach, hence thought of a one liner, but yeah, inline works too, let me update.

File: vhdbuilder/scripts/linux/ubuntu/tool_installs_ubuntu.sh (hunk 9 above)
Comment (alexeldeib) [reply vhdbuilder/scripts/linux/ubuntu/tool_installs_ubuntu.sh:157]: eh just git struggling

File: vhdbuilder/packer/vhd-image-builder-mariner-arm64.json [hunk 10]
```diff
@@ -458,7 +458,19 @@
     {
       "type": "shell",
       "inline": [
-        "sudo FEATURE_FLAGS={{user `feature_flags`}} BUILD_NUMBER={{user `build_number`}} BUILD_ID={{user `build_id`}} COMMIT={{user `commit`}} HYPERV_GENERATION={{user `hyperv_generation`}} CONTAINER_RUNTIME={{user `container_runtime`}} TELEPORTD_PLUGIN_DOWNLOAD_URL={{user `teleportd_plugin_download_url`}} ENABLE_FIPS={{user `enable_fips`}} SGX_INSTALL={{user `sgx_install`}} IMG_SKU={{user `img_sku`}} /bin/bash -ux /home/packer/pre-install-dependencies.sh",
+        "sudo FEATURE_FLAGS={{user `feature_flags`}} BUILD_NUMBER={{user `build_number`}} BUILD_ID={{user `build_id`}} COMMIT={{user `commit`}} HYPERV_GENERATION={{user `hyperv_generation`}} CONTAINER_RUNTIME={{user `container_runtime`}} TELEPORTD_PLUGIN_DOWNLOAD_URL={{user `teleportd_plugin_download_url`}} ENABLE_FIPS={{user `enable_fips`}} SGX_INSTALL={{user `sgx_install`}} IMG_SKU={{user `img_sku`}} /bin/bash -ux /home/packer/pre-install-dependencies.sh"
+      ]
+    },
+    {
+      "type": "shell",
```
Comment (alexeldeib) [inline vhdbuilder/packer/vhd-image-builder-mariner-arm64.json:465]: why not remove the latter 2x reboots, and also remove them from ubuntu + add this file for ubuntu...?

File: vhdbuilder/scripts/linux/ubuntu/tool_installs_ubuntu.sh (hunk 9 above)
Comment (anujmaheshwari1) [reply vhdbuilder/scripts/linux/ubuntu/tool_installs_ubuntu.sh:157]: yes lol

File: vhdbuilder/packer/pre-install-dependencies.sh (hunk 8 above)
Comment (anujmaheshwari1) [reply vhdbuilder/packer/pre-install-dependencies.sh]: addressed

File: vhdbuilder/packer/pre-install-dependencies.sh (hunk 7 above)
Comment (anujmaheshwari1) [reply vhdbuilder/packer/pre-install-dependencies.sh]: resolved.

File: vhdbuilder/packer/vhd-image-builder-mariner-arm64.json (hunk 10 above)
Comment (anujmaheshwari1) [reply vhdbuilder/packer/vhd-image-builder-mariner-arm64.json]: done

File: vhdbuilder/packer/pre-install-dependencies.sh [hunk 11]
```diff
@@ -48,11 +48,45 @@ rm -f /etc/cron.daily/logrotate
...
+  # so we just hold the kernel image packages for now on CVM.
+  # this still allows us base image and package updates on a weekly cadence.
+  if [[ "$IMG_SKU" != "20_04-lts-cvm" ]]; then
+    apt_get_update || exit $ERR_APT_UPDATE_TIMEOUT
+    apt_get_dist_upgrade || exit $ERR_APT_DIST_UPGRADE_TIMEOUT    
+  fi
+
+  if [[ "${ENABLE_FIPS,,}" == "true" ]]; then
+    # This is FIPS Install for Ubuntu, it purges non FIPS Kernel and attaches UA FIPS Updates
+    echo "Install FIPS for Ubuntu SKU"
```
Comment (alexeldeib) [inline vhdbuilder/packer/pre-install-dependencies.sh:80]: this could actually be mariner in both the comment and the echo, right?

File: vhdbuilder/packer/pre-install-dependencies.sh [hunk 12]
```diff
@@ -48,11 +48,45 @@ rm -f /etc/cron.daily/logrotate
...
+    installFIPS
+  fi
+
+  # Final step, if 1804 or FIPS, log ua status, detach UA and clean up
+  if [[ "${UBUNTU_RELEASE}" == "18.04" ]] || [[ "${ENABLE_FIPS,,}" == "true" ]]; then
```
Comment (alexeldeib) [inline vhdbuilder/packer/pre-install-dependencies.sh:85]: the second condition here needs to check ubuntu vs mariner

File: vhdbuilder/packer/pre-install-dependencies.sh [hunk 13]
```diff
@@ -48,11 +48,45 @@ rm -f /etc/cron.daily/logrotate
...
+# First handle Mariner + FIPS
+if [[ ${OS} == ${MARINER_OS_NAME} ]]; then
+  if [[ "${ENABLE_FIPS,,}" == "true" ]]; then
+    # This is FIPS install for Mariner and has nothing to do with Ubuntu Advantage
+    echo "Install FIPS for Mariner SKU"
+    installFIPS
+  fi
+else
+  # Handle FIPS and ESM for Ubuntu
+  if [[ "${UBUNTU_RELEASE}" == "18.04" ]] || [[ "${ENABLE_FIPS,,}" == "true" ]]; then
```
Comment (alexeldeib) [inline vhdbuilder/packer/pre-install-dependencies.sh:60]: the second condition here needs to check ubuntu vs mariner

Comment (alexeldeib) [reply vhdbuilder/packer/pre-install-dependencies.sh:60]: nvm, this is all in else block

File: vhdbuilder/packer/pre-install-dependencies.sh (hunk 12 above)
Comment (alexeldeib) [reply vhdbuilder/packer/pre-install-dependencies.sh:85]: nvm, this is all in else block

File: vhdbuilder/packer/pre-install-dependencies.sh (hunk 11 above)
Comment (alexeldeib) [reply vhdbuilder/packer/pre-install-dependencies.sh:80]: nvm, this is all in else block

**Review by alexeldeib — APPROVED** [review]

**Review by cameronmeissner — APPROVED** [review]

**Review by hbeberman — APPROVED** [review]

**Review by chengliangli0918 — APPROVED** [review]
